ray-patching.md

```diff
@@ -1,53 +1,116 @@
+* `git rebase -i tags/ray-<new_version>` (the `-i` is optional, but you can
+  use it to drop patches that you know where upstreamed even if git doesn't auto-recognize them)
+* Fix eventual conflicts, taking care to keep the original patch attribution
+  (if you make substantial changes, add a `Co-Authored-By: ...`)
+  * I.e. if git tells you it couldn't apply commit `deadbeef`, then, after fixing the conflicts, do
+  * `git add .`
+  * `git show deadbeef`
+  * `git commit --date="<date_of_deadbeef>" --author="<author_of_deadbeef>"` (commit message is kept as-is automatically while rebasing)
+  * `git rebase --continue`
+* `git format-patch tags/ray-<new_version>`
+* copy resulting patches over to recipe and update `meta.yaml`
+* remove all other patches in the `patches`  folder
+
 ## How to prepare or adapt patches for third party Ray components (using Redis as the example)
 
+The above process is complicated by the fact that ray has patches for third party
+dependencies (like redis) in its source code - in other words, we need to patch
+the patches.
+
+The process we just described then repeats for each dependency we need to patch
+more or less as-is, with the exception that we now need to check in the patches
+to third-party code into the ray source-code, and include _that_ in the ray-patches
+we check into the feedstock. For added fun, it may happen that upstream ray changes
+their patches from version to version in ways that conflict with ours.
+
+Since patch application within the ray build is done by bazel (and may be subject
+to destructive options like `-pN`, which changes the file hierarchy level the
+patches apply to), it's best - for single patches at least - to extend the
+ray patch directly to suit our purposes. This makes it also much clearer (well...)
+when reviewing the diff(-of-the-diff-of-the-diff) on the feedstock.
+
 ### Prepare sources
-1. Get Ray repository (using Ray 1.1.0 for example)
+1. Get Ray repository (using Ray 2.1.0 for example)
 ```sh
 git clone https://github.com/ray-project/ray.git
 cd ray
-git checkout -b ray-1.1.0-original ray-1.1.0
+git checkout -b ray-2.1.0-original ray-2.1.0
 # apply recipe patches
 git am /ray-packages-feedstock/recipe/patches/*
 cd ..
 ```
-2. Find out repository and version needed (see `ray/bazel/ray_deps_setup.bzl`), clone it and checkout (taking version 6.0.9 for example):
+2. Find out repository and version needed (see `ray/bazel/ray_deps_setup.bzl`), clone it and checkout (taking version 7.0.5 for example):
 ```sh
 git clone  https://github.com/redis/redis
 cd redis
-git checkout -b 6.0.9-original 6.0.9
+git checkout -b 7.0.5-patched 7.0.5
 ```
-3. Apply relevant upstream patches first (1.1.0 has only one patch) and commit them:
+3. Apply relevant upstream patches first (2.1.0 has only one patch) and commit them:
 ```sh
-patch -p0 < patch -p0 < ../ray/thirdparty/patches/redis-quiet.patch
+patch -p1 < patch -p1 < ../ray/thirdparty/patches/redis-quiet.patch
```

Comment:
I now forgot why the line is like this, why it has double `patch -p1` repeated?..

Reply:
I don't know, the double `patch` was there before already, I just adapted it to the new upstream situation of requiring -p1 (which matches git's default as well)

Reply:
```suggestion
patch -p1 < ../ray/thirdparty/patches/redis-quiet.patch
```
this looks suspicially like a typo, we should probably fix that